Extend `Dropdown::ListItem`

alex-ju opened the pull request:
### :pushpin: Summary

Extend `Hds::Dropdown::ListItem` with selectable items and the option for result counts to support the upcoming 'Filter' pattern.

### :hammer_and_wrench: Detailed description

- Add `Hds::Dropdown::ListItem::Checkmark` as a proxy for `Hds::Interactive` with a `@selected` argument (for context switching)
- Add `Hds::Dropdown::ListItem::Checkbox` as a proxy for `Hds::Form::Checkbox` (for filter)
- Add `Hds::Dropdown::ListItem::Radio` as a proxy for `Hds::Form::Radio` (for filter)
- Add `@count` argument to selectable list items (Checkmark, Checkbox, Radio) 

> **Note**
> Internal class name changes: `hds-dropdown-list-item--[variant]` → `hds-dropdown-list-item--variant-[variant]`
> - `hds-dropdown-list-item--copy-item` → `hds-dropdown-list-item--variant-copy-item`
> - `hds-dropdown-list-item--description` → `hds-dropdown-list-item--variant-description`
> - `hds-dropdown-list-item--generic` → `hds-dropdown-list-item--variant-generic`
> - `hds-dropdown-list-item--interactive` → `hds-dropdown-list-item--variant-interactive`
> - `hds-dropdown-list-item--separator` → `hds-dropdown-list-item--variant-separator`
> - `hds-dropdown-list-item--title` → `hds-dropdown-list-item--variant-title`
>
> These should not represent a breaking change as the consumers rely on the public API, but if test assertions are based on class names or extensions/overrides have been applied to these elements they may raise issues. For this reason, we flag this change as a **breaking change** to give consumers a heads-up.

[👉 Preview showcase](https://hds-showcase-git-alex-ju-dropdown-extension-hashicorp.vercel.app/components/dropdown)

Related work:
- [x] [Test the current API](https://github.com/hashicorp/cloud-ui/commit/d63bb0f6f93771f056d3e6c7efdb7b01fcd46a49) in context (`cloud-ui`s Vault insights seems like a good candidate)
- [x] Figure out if additional filter features (such as allowing result count in content items) should be part of this work or a follow-up PR – decided to group the selectable list item with the result count in this PR and leave search and actions for separate PRs so we can deliver the whole filter pattern work incrementally

<table>
<tr><th>Before</th><th>After</th></tr>
<tr><td>

![checkbox-before](https://user-images.githubusercontent.com/788096/218160757-1959618f-4ff0-4570-a733-99c4d1fb96f0.png)


</td><td>

![checkbox-after](https://user-images.githubusercontent.com/788096/218160770-659684df-d330-4abd-9d51-34393ff0461c.png)


</td></tr>
<tr><td>

![radio-before](https://user-images.githubusercontent.com/788096/218160788-ecf8693d-57f6-4b6a-bf05-4dfbc072792d.png)


</td><td>

![radio-after](https://user-images.githubusercontent.com/788096/218160803-d8452cf8-e509-4717-8dd6-26d024c6a282.png)


</td></tr>
</table>


### :link: External links

<!-- Issues, RFC, etc. -->
Jira ticket: Part of [HDS-1109](https://hashicorp.atlassian.net/browse/HDS-1109)
Figma file: https://www.figma.com/file/Kgc8BgB8jGuJXISndgTrD1/Dropdown?node-id=27919%3A58777&t=4MEZaQEQPfhRuVTW-4

***

### 👀 Reviewer's checklist:

- [x] +1 Percy if applicable
- [x] Confirm that PR has a changelog update via [Changesets](https://github.com/changesets/changesets) if needed

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.


[HDS-1109]: https://hashicorp.atlassian.net/browse/HDS-1109?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

alex-ju commented on packages/components/addon/components/hds/dropdown/index.hbs:
```diff
@@ -8,14 +8,17 @@
   </:toggle>
   <:content as |c|>
-    <ul class={{this.listClassNames}} {{style width=@width}}>
+    <ul class={{this.listClassNames}} {{style width=@width}} role="listbox">
```
This role would need to be applied only when using the `Checkmark` contextual component. This raises the usual challenge of a parent component knowing about children.

MelSumner replied:
Have you tried using a conditional here? 

```
role={{if hasCheckbox "listbox" null }}
```

Ember won't render an attribute at all if the value is `null`.

alex-ju replied:
yes, using a boolean here would be an option, just have to make sure we either document it well or we try to determine it automatically within the component

alex-ju commented on packages/components/app/styles/components/dropdown.scss:
```diff
@@ -408,8 +416,82 @@ $hds-dropdown-toggle-border-radius: 5px;
+  background-size: var(--token-form-select-background-image-size) var(--token-form-select-background-image-size);
+
+  .hds-dropdown-list-item__interactive {
+    color: var(--token-color-foreground-action-active);
```
```suggestion
    color: var(--token-color-foreground-action);
```

didoo commented on packages/components/app/styles/components/dropdown.scss:
```diff
@@ -408,8 +416,82 @@ $hds-dropdown-toggle-border-radius: 5px;
+  background-position: top 10px right 10px;
+  background-size: var(--token-form-select-background-image-size) var(--token-form-select-background-image-size);
+
+  .hds-dropdown-list-item__interactive {
```
maybe it's me, but may be that this rule is not applied? I don't see it in the styles when I inspect it in the browser (but may be doing something wrong).

alex-ju replied:
yes, it wasn't applied, because of lack of specificity – adjusted now

didoo commented on packages/components/app/styles/components/dropdown.scss:
```diff
@@ -408,8 +416,82 @@ $hds-dropdown-toggle-border-radius: 5px;
+  padding: 8px 16px;
+
+  .hds-form-field--layout-flag {
+    flex: 1;
```
Probably I would not override this, but use a wrapping container (or rely on the `auto` margin of the elements on the right, if possible). Not a blocker though.

didoo commented on packages/components/app/styles/components/dropdown.scss:
```diff
@@ -408,8 +416,82 @@ $hds-dropdown-toggle-border-radius: 5px;
+  }
+
+  .hds-form-label {
+    font-weight: normal;
```
In the designs in Figma, I see the color is not `form/label` but `foreground/primary`. Can you check with @heatherlarsen which one we should use?

heatherlarsen replied:
`foreground/primary` is correct. This is to keep the color of the label when it is within the menu, consistent with the other dropdown list item variants, which use `foreground/primary`.

didoo replied:
In this case, another reason (not strong, just a possibility) to considering using the `Base` form element, and not the `Field`, like it's been done in the `Pagination` component (and in that case, it's likely you will not need to override the `hds-form-field--layout-flag` flex).

alex-ju replied:
good spot @didoo and thank you for confirming @heatherlarsen; will try with the `Base` element the to minimise the overrides

alex-ju replied:
updated to [use the form base element instead of form field](https://github.com/hashicorp/design-system/pull/1185/commits/4aefe474463b81fe1b281b8b9a85a3f7c9029281)

didoo reviewed (commented):
@alex-ju did a first high-level pass at the code and added a few comments.

(didn't look at the tests, for now)

heatherlarsen commented on packages/components/addon/components/hds/dropdown/list-item/checkbox.hbs:
```diff
@@ -0,0 +1,8 @@
+    <F.Label>{{yield}}</F.Label>
+  </Hds::Form::Checkbox::Field>
+  {{#if @resultCount}}
+    <Hds::BadgeCount class="hds-dropdown-list-item__count" @size="small" @type="filled" @text={{@resultCount}} />
```
This is my fault. I used the BadgeCount component as a starting point in the explorations because we were looking at using them directly and for this variation of the design, I just detached and overrode the values. I'll make sure the built component is just plain text with a container. It makes sense to me that the implemented version does not use the BadgeCount component.

alex-ju replied:
thank you both! will update to use a plain element then

alex-ju replied:
[We're now using a plain element for the result count](https://github.com/hashicorp/design-system/pull/1185/commits/d3d7cc80906c0163b0d39d66f22f6b0932068704) (replaced `BadgeCount` with `<span>` and applied similar style)

alex-ju commented on packages/components/app/styles/components/dropdown.scss:
```diff
@@ -408,8 +416,82 @@ $hds-dropdown-toggle-border-radius: 5px;
+  }
+}
+
+.hds-dropdown-list-item--checkmark-selected {
```
I went with the background image to preserve the result count position/alignment for both selected and unselected items; I've updated the code to [use FlightIcon for checkmark](https://github.com/hashicorp/design-system/pull/1185/commits/21f97281478d34c1a04b925e7754c2194385703b) for selected items, and rely on padding for unselected ones.

alex-ju commented on packages/components/app/styles/components/dropdown.scss:
```diff
@@ -208,18 +209,23 @@ $hds-dropdown-toggle-border-radius: 5px;
-.hds-dropdown-list-item--interactive {
+// HDS::DROPDOWN::LIST-ITEM::INTERACTIVE & HDS::DROPDOWN::LIST-ITEM::CHECKMARK
+.hds-dropdown-list-item--interactive,
+.hds-dropdown-list-item--checkmark {
```
we still need a `--checkmark` modifier operating some overrides, so instead of composing from two modifiers (`--interactive` and `--checkmark` I thought I may use the same properties block. Some of these properties could potentially be abstracted in a mixin to simplify the code a bit.

didoo replied:
👍

alex-ju commented on packages/components/addon/components/hds/dropdown/list-item/checkmark.hbs:
```diff
@@ -0,0 +1,24 @@
+{{! template-lint-disable require-context-role require-presentational-children }}
+<li class={{this.classNames}}>
+  <Hds::Interactive
```
`cloud-ui` doesn't do URL updating on click, but on a separate 'apply filter' button, but HashiDocs, for example, does it on click, and I imagine other products might take this approach (as I understand we're aiming to support both, by design).

didoo replied:
👍

didoo commented on packages/components/app/styles/components/dropdown.scss:
```diff
@@ -408,8 +416,86 @@ $hds-dropdown-toggle-border-radius: 5px;
+  a.hds-dropdown-list-item__interactive,
+  button.hds-dropdown-list-item__interactive {
+    // leave room for the checkmark to align the result count on non-selected items
+    padding-right: 25px; // 9px default + 16px for the 'check' FlightIcon
```
Maybe you could use a parent container, with its own intrinsic size, and then conditionally render the icon inside it, so that you don't need to use extra padding (and calculations) to achieve the desired layout

alex-ju replied:
Followed the suggestion and refactored to use a fixed-width container, that conditionally renders the checkmark icon in [35201dd](https://github.com/hashicorp/design-system/pull/1185/commits/35201dd1eb17d523f19e16967eeac30e8dfe3875)

didoo commented on packages/components/app/styles/components/dropdown.scss:
```diff
@@ -408,8 +416,86 @@ $hds-dropdown-toggle-border-radius: 5px;
+
+.hds-dropdown-list-item__checkmark {
+  align-self: flex-start;
+  margin-top: 0.125rem;
```
if you use a container you don't need this padding anymore (you can set the height directly to the desired value, which is `20px`, the line-height of the main text)

didoo commented on packages/components/app/styles/components/dropdown.scss:
```diff
@@ -408,8 +416,86 @@ $hds-dropdown-toggle-border-radius: 5px;
+  align-self: flex-start;
+  padding: 0.125rem 0.5rem;
+  color: var(--token-color-foreground-faint);
+  line-height: 1.2308; // 16px
```
maybe instead of doing this double trick of setting a padding + a line height, you can be more declarative and set the line-height to `20px` (the line-height of the main text) and no vertical padding (adding a comment explaining why this 20px value)

alex-ju replied:
that makes sense, especially now when the checkmark icon also uses this value – updated in [3991ef2](https://github.com/hashicorp/design-system/pull/1185/commits/3991ef22579e58c852e1f7cbadc57b66b81487a9)

didoo commented on packages/components/tests/integration/components/hds/dropdown/list-item/checkmark-test.js:
```diff
@@ -0,0 +1,92 @@
+      assert
+        .dom('.hds-dropdown-list-item')
+        .hasClass('hds-dropdown-list-item--checkmark')
+        .hasClass('hds-dropdown-list-item--color-action');
```
I've mixed feeling about this. Isn't this class mainly presentational?
```suggestion
```

didoo commented on packages/components/tests/integration/components/hds/dropdown/list-item/checkbox-test.js:
```diff
@@ -0,0 +1,66 @@
+import { hbs } from 'ember-cli-htmlbars';
+
+module(
+  'Integration | Component | hds/dropdown/list-item/checkbox',
```
I see below we are testing the "selected" state of the "checkmark" which made me think: should we test also the "checked" state of the `list-item/checkbox`? (and radio too?)

didoo replied:
also, should we test the `getId` functionality too? (the label and control having the correct ID?)

alex-ju replied:
amended the tests in [121b639](https://github.com/hashicorp/design-system/pull/1185/commits/121b6396c33a83bca71f09c1805ad12f7de8fb43) to capture these features

didoo reviewed (commented):
left a few comments/suggestions

alex-ju commented:
@didoo, @heatherlarsen, @MelSumner I rebase this to main to align with the new Showcase – this is good for another review

didoo commented on packages/components/app/styles/components/dropdown.scss:
```diff
@@ -413,8 +421,88 @@ $hds-dropdown-toggle-border-radius: 5px;
+  }
+
+  .hds-dropdown-list-item__interactive-text {
+    flex: 1;
```
have you considered moving this declaration directly to the `.hds-dropdown-list-item__interactive-text` instead of here? I _don't think_ there would be any unintended side effects, and would avoid having too many specific/custom declarations around (even the overrides of the `focus/active` states above make me nervous: was it _really_ necessary?).

alex-ju replied:
Indeed this property can be applied to the interactive text with no side effects.
Disabling the background color for focus/active states was needed to allow the use of badges inside the checkmark list item while [preserving a compliant contrast](https://www.figma.com/file/Kgc8BgB8jGuJXISndgTrD1/Dropdown?node-id=28361%3A64942&t=3jf3TMGNrOAbdpgZ-4).

didoo commented on packages/components/app/styles/components/dropdown.scss:
```diff
@@ -413,8 +421,88 @@ $hds-dropdown-toggle-border-radius: 5px;
+.hds-dropdown-list-item--checkmark-selected {
+  a.hds-dropdown-list-item__interactive,
+  button.hds-dropdown-list-item__interactive {
+    padding-right: 9px; // reset to default
```
I _think_ you don't need this anymore, can you double check?
```suggestion
```

alex-ju replied:
Indeed, this was leftover – removed it

didoo commented on packages/components/app/styles/components/dropdown.scss:
```diff
@@ -413,8 +421,88 @@ $hds-dropdown-toggle-border-radius: 5px;
+}
+
+.hds-dropdown-list-item__checkmark-icon {
+  height: inherit;
```
too much "magic" for me here :) - I would prefer to have the parent as flexbox and use that to align the icon, it would be more explicit

alex-ju replied:
Update to make use of flex model

didoo commented on packages/components/app/styles/components/dropdown.scss:
```diff
@@ -413,8 +421,88 @@ $hds-dropdown-toggle-border-radius: 5px;
+
+  .hds-dropdown-list-item__label {
+    flex-grow: 1;
+    width: fit-content;
```
is this needed? I tried to turn it on/off in the devtools panel but I can't see a difference in the showcase

alex-ju replied:
not needed anymore in the new flex layout – removed it

didoo commented on packages/components/app/styles/components/dropdown.scss:
```diff
@@ -413,8 +421,88 @@ $hds-dropdown-toggle-border-radius: 5px;
+.hds-dropdown-list-item--checkmark,
+.hds-dropdown-list-item--checkbox,
+.hds-dropdown-list-item--radio {
+  // align any `Hds::Badge` within a selectable list item to match baseline
```
👏👏👏

didoo commented on packages/components/tests/dummy/app/templates/components/dropdown.hbs:
```diff
@@ -306,4 +306,298 @@
 
+  <Shw::Text::H3>Checkmark</Shw::Text::H3>
+  <Shw::Flex as |SF|>
+    <SF.Item>
```
I would add `@label` declarations to these (and other) items, is really hard to understand what they "showcase"

alex-ju replied:
Good suggestion – added

didoo commented on packages/components/tests/dummy/app/templates/components/dropdown.hbs:
```diff
@@ -306,4 +306,298 @@
+      <ul class="hds-dropdown-list" role="listbox">
+        {{#each @model.ITEM_STATES as |state|}}
+          <Hds::Dropdown::ListItem::Checkmark mock-state-value={{state}}>
+            <Shw::Placeholder @text="custom content" @width="150" @height="20" @background="#eee" />
```
super-nitpick: if you use `144` instead of `150`, the items will be better aligned (and will make easier for the eye to compare them)

alex-ju replied:
Thanks, yes, that was the initial intention, but haven't updated the values after the layout refactor.

heatherlarsen commented:
@alex-ju I just looked through the showcase...this is great ✨ 

[visual bug] I tested the updates in the showcase and noticed that there's still a light blue background that appears when clicking on the list items. (example: [screencast-calendar.google.com-2023.02.27-14_00_37.webm](https://user-images.githubusercontent.com/8553306/221695941-cc6e6001-82df-4459-b1e6-a5f05be2b790.webm)) Can that be removed please?

[question] Can the custom content versions render a count as well?

alex-ju commented:
> @alex-ju I just looked through the showcase...this is great ✨
> 
> [visual bug] I tested the updates in the showcase and noticed that there's still a light blue background that appears when clicking on the list items. (example: [screencast-calendar.google.com-2023.02.27-14_00_37.webm](https://user-images.githubusercontent.com/8553306/221695941-cc6e6001-82df-4459-b1e6-a5f05be2b790.webm)) Can that be removed please?

Good spot! There was a declaration left that caused that behavior – removed it now.

> [question] Can the custom content versions render a count as well?

Yes, they can. I updated the custom content examples to show the count

heatherlarsen commented:
@alex-ju Is there an option to add an icon to the checkmark, checkbox, and radio list items? Apologies that this wasn't in the Figma file; I must've forgotten to pull those in from the graveyard, but I've added that back into the file here:  https://www.figma.com/file/Kgc8BgB8jGuJXISndgTrD1/Dropdown?node-id=29345%3A66370&t=mlu3WJnbpLdrRqAE-4

I noticed the checkmark icon isn't optically aligned with the text. In Figma, I had to add a top padding of 2px above each icon. It's definitely hacky, but it fixes most of the imbalance issues. Let me know if you want to chat more about this, though.

alex-ju commented:
> @alex-ju Is there an option to add an icon to the checkmark, checkbox, and radio list items? Apologies that this wasn't in the Figma file; I must've forgotten to pull those in from the graveyard, but I've added that back into the file here: https://www.figma.com/file/Kgc8BgB8jGuJXISndgTrD1/Dropdown?node-id=29345%3A66370&t=mlu3WJnbpLdrRqAE-4

Noted – I'll make sure they can render icons as well

> I noticed the checkmark icon isn't optically aligned with the text. In Figma, I had to add a top padding of 2px above each icon. It's definitely hacky, but it fixes most of the imbalance issues. Let me know if you want to chat more about this, though.

Hmm.. I remember needing to add such padding, but then I updated the approach to [use an extra container](https://github.com/hashicorp/design-system/pull/1185#discussion_r1111809166) ensuring alignment with baseline and cap height/ascender. I'm wondering if I'm missing something here 🤔  (see screen capture below)

<img width="434" alt="Screenshot 2023-03-02 at 14 22 28" src="https://user-images.githubusercontent.com/788096/222456303-f9a57711-ab80-42c9-a302-704ca55d9628.png">

heatherlarsen reviewed (changes requested):
@alex-ju Ahh, you are correct; thank you! Maybe I was looking at an old link 🤔 

I've been able to go through the showcase more thoroughly for design alignment and found two more minor things: 

- there doesn't appear to be a gap between the text and the checkmark icon, so for longer strings, the text ends up bumping up right next to the icon. Is it possible to have an 8px gap between the text and the icon?
- because the checkmark icon is placed in a 20x20 container, there's an extra 4px to the right of the icon, which results in an optically larger padding on the right. The padding on the right should be a total of 10px, it currently looks like it's closer to 14px. Can the size of the container be reduced to 16px wide?

Let me know if we need to discuss these further. Thanks!

alex-ju commented:
> @alex-ju Ahh, you are correct; thank you! Maybe I was looking at an old link 🤔
> 
> I've been able to go through the showcase more thoroughly for design alignment and found two more minor things:
> 
> * there doesn't appear to be a gap between the text and the checkmark icon, so for longer strings, the text ends up bumping up right next to the icon. Is it possible to have an 8px gap between the text and the icon?
> * because the checkmark icon is placed in a 20x20 container, there's an extra 4px to the right of the icon, which results in an optically larger padding on the right. The padding on the right should be a total of 10px, it currently looks like it's closer to 14px. Can the size of the container be reduced to 16px wide?
> 
> Let me know if we need to discuss these further. Thanks!

Both are good spots @heatherlarsen! I made sure we have a consistent 8px spacing between the elements of a list item (icon, text, count, checkmark) and have updated the checkmark width to 16px to avoid any extra space there.

alex-ju commented:
> more comments :)

@didoo I have hopefully addressed the feedback from previous rounds of review at the time.

> question: do you remember we discussed the changes to the "modifiers" classes for the "list-item" variants (from `hds-dropdown-list-item--checkmark` to `hds-dropdown-list-item--variant-checkmark` for example, to avoid confusion with `hds-dropdown-list-item__checkmark`)? since it's a breaking change, I am OK to do it in a separate PR (so this is not a blocker) but let's not forget to do it. if instead you prefer to do it directly in this one, no problems for me as well (in which case you have to update the changeset file)

Indeed, I remember the conversation and have updated the modifier classes here as we're introducing some more modifiers; have also updated the changelog to reflect this change and marked it as a major (although technically it is not, we know it will cause breaking changes).

However, since the last review, a few more changes were introduced. I kept the git history untouched to help you pick up from where you left off, although may not be of much value if reviewed as a whole.

@heatherlarsen I have hopefully addressed your feedback too, including the updated vertical alignment. If you spot anything off please leave comments as necessary.

heatherlarsen reviewed (approved):
This looks great, @alex-ju! I don't see anything else that needs adjusting ✨ :shipit:

didoo commented on .changeset/olive-avocados-attack.md:
```diff
@@ -0,0 +1,6 @@
+---
+
+Allow `Checkmark`, `Checkbox` and `Radio` as `Hds::Dropdown::ListItem`s
+Rename `Hds::Dropdown::ListItem` internal CSS classes
```
do you think it would make sense to add here a note/comment that explains that _this_ is the breaking change that causes the version bump to be major?

alex-ju replied:
[Updated the changelog entry](https://github.com/hashicorp/design-system/pull/1185/commits/fcf80f8b846bf615aa648335793e9a7d39d78c60) to list the actual changes and the potential effects.

didoo reviewed (commented):
sorry, premature submit 🤦‍♂️

didoo commented on packages/components/tests/dummy/app/templates/components/dropdown.hbs:
```diff
@@ -306,4 +306,358 @@
+      </Hds::Dropdown>
+    </SF.Item>
+  </Shw::Flex>
+
```
I would probably add level-1 and level-2 separators (`<Shw::Divider />`, `<Shw::Divider @level={{2}} />`) between the different types examples, to increase readability/scan.
See for example how it's done in the [Pagination](https://hds-showcase.vercel.app/components/pagination) or [Form Base elements](https://hds-showcase.vercel.app/components/form/base-elements) showcases.

alex-ju replied:
Added the level 2 dividers – looks much better with them!

didoo commented on packages/components/tests/integration/components/hds/dropdown/index-test.js:
```diff
@@ -119,4 +119,27 @@ module('Integration | Component | hds/dropdown/index', function (hooks) {
+  test('it should render a list of items with a `listbox` role if selectable items are passed in', async function (assert) {
+    await render(hbs`
+      <Hds::Dropdown id="test-dropdown" as |dd|>
+      <dd.ToggleButton @text="toggle button" id="test-toggle-button" />
+      <dd.Checkmark>Checkmark</dd.Checkmark>
```
```suggestion
        <dd.ToggleButton @text="toggle button" id="test-toggle-button" />
        <dd.Checkmark>Checkmark</dd.Checkmark>
```

didoo reviewed (approved):
All looking (very) good to me. 👍

For the merge and release, let's sync: we probably want to merge (and release) the `SideNav` elements, which are not breaking, so we can start to adopt them, and then release these (breaking) ones, which need to be followed by one or more PRs in Cloud UI and potentially other consumers (see https://cs.github.com/?scope=org%3Ahashicorp&scopeName=hashicorp&q=hds-dropdown-list-item)

MelSumner commented on packages/components/tests/integration/components/hds/dropdown/index-test.js:
```diff
@@ -119,4 +119,27 @@ module('Integration | Component | hds/dropdown/index', function (hooks) {
+
+  // ACCESSIBILITY
+
+  test('it should render a list of items without a role if no selectable items are passed in', async function (assert) {
```
question: I don't really understand this test, can you explain? Lists show up in the rotor and as such need an accessible name even if they don't have interactive elements.

alex-ju replied:
> it should render a list of items without a role if no selectable items are passed in

now that I read it, I can see how the double negation doesn't help

what I'm testing here is that _if_ we don't have selectable items (checkbox, radio), we don't add `role="listbox"` to the list, relying on the implicit role for an unordered list, which is "list" (in this scenario, a screen reader would read: _my item, button, group, 5 items_)

in the following test, we check that we add `role="listbox"` if there are selectable items (in this scenario, a screen reader would read: _my item, not selected, 1 of 5_)

I'm trying to make sure the `role` is only explicitly applied to the list if it has selectable items so that users are told about the state of each item when navigating through them

alex-ju commented:
@didoo while testing this change in `cloud-ui` I realized that if we change the variant class names, it will also require a new `structure` release for the [override selectors](https://github.com/hashicorp/structure/blob/main/packages/pds-ember/app/styles/pds/_overrides.scss#L8).
